core/src/main/java/hudson/model/User.java

```diff
@@ -536,12 +540,45 @@ public synchronized void doSubmitDescription( StaplerRequest req, StaplerRespons
 
     /**
      * Gets the {@link User} object by its id or full name.
+     *
+     * Creates a user on-demand.
+     *
+     * <p>
      * Use {@link #getById} when you know you have an ID.
+     * In this method Jenkins will try to resolve the {@link User} by full name with help of various
+     * {@link hudson.tasks.UserNameResolver}.
+     * This is slow (see JENKINS-23281).
+     *
+     * @deprecated This method is deprecated, because it causes unexpected {@link User} creation
+     *             by API usage code and causes performance degradation of used to retrieve users by ID.
+     *             Use {@link #getById} when you know you have an ID.
+     *             Otherwise use {@link #getOrCreateByIdOrFullName(String)} or {@link #get(String, boolean, Map)}.
      */
+    @Deprecated
```

Comment:
What about deprecating `get(String, boolean, Map)` also? Or do you think that overload is weird enough that people would not even consider calling it until they had read the Javadoc and noticed the “or full name” part?

Reply:
Well, I definitely have plugins where I need to refactor that (ownership & Co). Maybe it also needs to be deprecated, but right now there is no full replacement assuming somebody really uses `context`